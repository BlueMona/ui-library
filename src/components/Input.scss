.p-input {
  display: flex;
  flex-grow: 1;

  position: relative;
  height: auto;
  width: 100%;

  input,
  textarea {
    border: 1px solid $input-border;
    border-radius: 4px;

    &::placeholder {
      color: $text-dark-hint;
    }
  }

  input {
    flex-grow: 1;
    height: 48px;
    max-width: 100%;
    padding: 0 $padding-small 0;
    margin-top: 0;
    transition: 150ms;
  }

  textarea {
    flex-grow: 1;
    padding: 12px $padding-small $padding-small;
  }

  .label {
    display: flex;
    align-items: center;
    height: 16px;
    position: absolute;
    pointer-events: none;
    top: -8px;
    left: $padding-tiny;
    padding: 0 $padding-tiny;
    transition: top 150ms, font-size 150ms, color 150ms;
    color: $text-dark-secondary;
    background-color: $background-root;
    font-size: 12px;
  }

  .hint {
    @extend %hint-error-label;
    position: absolute;
    pointer-events: none;
    top: 28px;
    padding-left: $padding-small;

    transition: opacity 150ms;
    color: $text-dark-hint;
    opacity: 0;

    &.visible {
      opacity: 1;
    }
  }

  // Hover state
  input:hover {
    border-color: $black;
  }

  // Focus state
  &.focused {
    input {
      outline: 0;
      border-color: $peerio-blue;
    }

    .error,
    + .helper-text {
      color: $peerio-blue;
    }
  }

  // Error state
  &.has-error {
    input {
      border-color: $color-error;
    }

    .error {
      @extend %hint-error-label;
      position: absolute;
      width: 100%;
      top: 100%;
      color: $color-error;
      opacity: 1;
    }
  }

  // Close button
  .clear-button {
    position: absolute;
    right: 0;
    top: 50%;
    margin-top: -18px; // half icon height
  }
}
