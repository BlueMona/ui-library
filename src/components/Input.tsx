import React from 'react';
import { action, observable } from 'mobx';
import { observer } from 'mobx-react';
import css from 'classnames';

import { Button } from './Button';
import { MaterialIcon } from './MaterialIcon';

interface BaseInputProps {
  className?: string;
  label?: string;
  error?: string;
  hint?: string;

  // Standard HTML input props
  autoFocus?: boolean;
  maxLength?: number;
  placeholder?: string;

  value?: string;

  // React props
  onChange?: (val: string) => void;
  onClear?: () => void;
  onFocus?: () => void;
  onBlur?: () => void;
  onKeyUp?: React.KeyboardEventHandler<HTMLInputElement | HTMLTextAreaElement>;
  onKeyDown?: React.KeyboardEventHandler<
    HTMLInputElement | HTMLTextAreaElement
  >;
  onKeyPress?: React.KeyboardEventHandler<
    HTMLInputElement | HTMLTextAreaElement
  >;
}

interface TextAreaInputProps {
  multiline: true;
  innerRef?: React.Ref<HTMLTextAreaElement>;
}

interface InputInputProps {
  multiline?: false;
  type?: 'text' | 'password';
  readOnly?: boolean;
  disabled?: boolean;
  innerRef?: React.Ref<HTMLInputElement>;
}

export type InputProps = BaseInputProps &
  (TextAreaInputProps | InputInputProps);

@observer
export class Input extends React.Component<InputProps> {
  @observable
  isFocused = false;
  @observable
  inputRef: HTMLInputElement | HTMLTextAreaElement | null = null;

  handleChange = (
    ev: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    if (!this.props.onChange) return;
    this.props.onChange(ev.currentTarget.value);
  };

  clearInput = () => {
    if (this.props.onClear) this.props.onClear();
  };

  @action.bound
  handleFocus() {
    if (this.props.onFocus) this.props.onFocus();
    this.isFocused = true;
  }

  @action.bound
  handleBlur() {
    if (this.props.onBlur) this.props.onBlur();
    this.isFocused = false;
  }

  @action.bound
  setRef(ref: HTMLInputElement | HTMLTextAreaElement | null) {
    if (ref) {
      this.inputRef = ref;
      if (this.props.autoFocus) {
        this.focus();
      }
    }
  }

  @action.bound
  focus() {
    if (this.inputRef) {
      this.inputRef.focus();
      this.handleFocus();
    }
  }

  @action.bound
  blur() {
    if (this.inputRef) {
      this.inputRef.blur();
      this.handleBlur();
    }
  }

  render() {
    return (
      <div
        className={css('p-input', this.props.className, {
          'has-label': !!this.props.label,
          'has-error': !!this.props.error,
          focused: this.isFocused
        })}
      >
        {this.props.label ? (
          <div className={css('label')}>{this.props.label}</div>
        ) : null}

        {this.props.multiline ? (
          <textarea
            placeholder={this.props.placeholder}
            value={this.props.value}
            maxLength={this.props.maxLength}
            onChange={this.props.onChange ? this.handleChange : undefined}
            onKeyPress={this.props.onKeyPress}
            onKeyDown={this.props.onKeyDown}
            onKeyUp={this.props.onKeyUp}
            onBlur={this.handleBlur}
            onFocus={this.handleFocus}
            ref={this.props.innerRef || this.setRef}
          />
        ) : (
          <input
            placeholder={this.props.placeholder}
            value={this.props.value}
            maxLength={this.props.maxLength}
            onChange={this.props.onChange ? this.handleChange : undefined}
            onKeyPress={this.props.onKeyPress}
            onKeyDown={this.props.onKeyDown}
            onKeyUp={this.props.onKeyUp}
            onBlur={this.handleBlur}
            onFocus={this.handleFocus}
            type={this.props.type || 'text'}
            readOnly={this.props.readOnly}
            disabled={this.props.disabled}
            ref={this.props.innerRef || this.setRef}
          />
        )}

        {!this.props.multiline && !!this.props.value && !!this.props.onClear ? (
          <Button
            className="clear-button"
            icon="close"
            onClick={this.clearInput}
          />
        ) : null}

        <div
          className={css('hint-or-error', {
            error: !!this.props.error,
            hint: !!this.props.hint,
            visible: !!this.props.error || (!!this.props.hint && this.isFocused)
          })}
        >
          {this.props.error ? (
            <React.Fragment>
              <MaterialIcon icon="error_outline" />
              {this.props.error}
            </React.Fragment>
          ) : this.props.hint ? (
            this.props.hint
          ) : null}
        </div>
      </div>
    );
  }
}
