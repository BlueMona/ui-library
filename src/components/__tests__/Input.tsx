import React from 'react';
import { shallow, mount } from 'enzyme';

import { Input } from '../Input';

describe('<Input>', () => {
  it('should match the snapshot', () => {
    expect(shallow(<Input value="Hello input!" />)).toMatchSnapshot();
  })

  it('should contain a dom <input> if multiline is not specified', () => {
    expect(shallow(<Input />).find('input')).toExist();
  })

  it('should contain a dom <textarea> if multiline is specified', () => {
    expect(shallow(<Input multiline />).find('textarea')).toExist();
  })

  it('should accept an innerRef of the correct type and assign it (input)', () => {
    const ref = React.createRef<HTMLInputElement>();
    mount(<Input innerRef={ref} />);
    expect(ref.current).not.toBeNull();
  })

  it('should accept an innerRef of the correct type and assign it (textarea)', () => {
    const ref = React.createRef<HTMLTextAreaElement>();
    mount(<Input multiline innerRef={ref} />);
    expect(ref.current).not.toBeNull();
  })
})