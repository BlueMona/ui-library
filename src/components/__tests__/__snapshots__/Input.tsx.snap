// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input> should match the snapshot 1`] = `
<div
  className="p-input"
>
  <div
    className="input-container"
  >
    <input
      onBlur={[Function]}
      onFocus={[Function]}
      type="text"
      value="Hello input!"
    />
  </div>
  <div
    className="hint-or-error"
  />
</div>
`;
